File: python/player.py
"""
Representations for Players
"""
def check_playable(card, played_cards):
    """
    Given a list of played cards, tells whether the given
    card is playable or not.
    """
    count = sum(map(lambda card: card.value, played_cards))
    if count + card.value > 31:
        return False
    return True

class Player:
    """
    Represents a player in the game. The base player just makes
    random decisions for throw_to_crib() and play_card().
    """
    def __init__(self, name):
        self.hand_cards = []
        self.points = 0
        self.name = name

    def __str__(self):
        return self.name + ' ' + str(self.hand_cards) + ' (' + str(self.points) + ')'

    def throw_to_crib(self, _):
        """
        Takes 2 cards out of self.hand_cards and returns them
        as a list.
        """
        return [self.hand_cards.pop(), self.hand_cards.pop()]

    def play_card(self, played_cards):
        """
        Removes and returns a card from self.hand_cards if there
        is a playable card. Otherwise, returns None (go).
        """
        for card in self.hand_cards:
            if not check_playable(card, played_cards):
                continue
            self.hand_cards.remove(card)
            return card
        return None

class HumanPlayer(Player):
    """
    Player that takes decisions from stdin.
    """
    def select_card(self):
        """
        Allows a card to be selected from the player's hand.
        """
        print('Enter a number to select the corresponding card.')
        print(self.hand_cards)
        for arrow in [True, False]:
            print(' ', end='')
            for (i, card) in enumerate(self.hand_cards):
                if card.rank == 10:
                    print(' ', end='')
                print(('^' if arrow else str(i + 1)) + '   ', end='')
            print()
        index = -1
        while index < 0:
            try:
                index = int(input())
                if index < 1 or index > len(self.hand_cards):
                    index = -1
                    raise ValueError()
            except ValueError:
                print('Please enter a number between 1 and ' + str(len(self.hand_cards)))
        selected_card = self.hand_cards[index - 1]
        self.hand_cards.remove(selected_card)
        return selected_card

    def throw_to_crib(self, dealer):
        cards_to_throw = []
        print('Throw 2 cards into your' + (' ' if dealer else " opponent's ") + 'crib.')
        for card_name in ['first', 'second']:
            print()
            print('Throw the ' + card_name + ' card.')
            card_to_throw = self.select_card()
            cards_to_throw.append(card_to_throw)
        print()
        return cards_to_throw

    def play_card(self, played_cards):
        for card in self.hand_cards:
            if check_playable(card, played_cards):
                print()
                print('Play a card.')
                selected_card = self.select_card()
                if check_playable(selected_card, played_cards):
                    return selected_card
                print(str(selected_card) + ' is not playable. Select a different card.')
                self.hand_cards.append(selected_card)
                return self.play_card(played_cards)
        print('No playable cards left.')
        return None

File: ts/player.ts
/**
 * Representations for Players
 */
import { Card, Suit } from './cards';
import prompts from 'prompts';

/**
 * @brief Given a list of played cards, tells whether the given card is
 * playable or not.
 */
const checkPlayable = (card: Card, playedCards: Card[]) => {
    const count = playedCards
        .map((card) => {
            return card.value;
        })
        .reduce((total, next) => {
            return total + next;
        }, 0);
    return count + card.value <= 31;
};

/**
 * @brief Represents a player in the game.
 *
 * The base player just makes random decisions for throwToCrib() and playCard().
 */
export class Player {
    private points;
    protected handCards: Card[];
    public constructor(private name: string) {
        this.points = 0;
        this.handCards = [
            new Card(Suit.HEARTS, 1),
            new Card(Suit.CLUBS, 4),
            new Card(Suit.SPADES, 2),
            new Card(Suit.SPADES, 10),
            new Card(Suit.SPADES, 13),
        ];
    }

    public readonly getPoints = () => {
        return this.points;
    };

    public readonly addPoints = (points: number) => {
        this.points += points;
    };

    public readonly getName = () => {
        return this.name;
    };

    public readonly toString = () => {
        return `${this.name} ${this.handCards} (${this.points})`;
    };

    /**
     * @brief Takes 2 cards out of self.hand_cards and returns them as a list.
     */
    public readonly throwToCrib = async (_: boolean) => {
        if (this.handCards.length < 2) {
            throw new Error(
                'Fatal error: Hand contained fewer than 2 cards when throwToCrib was called.',
            );
        }
        return [this.handCards.pop(), this.handCards.pop()] as Card[];
    };

    /**
     * @brief Removes and returns a card from self.hand_cards if there is a
     * playable card. Otherwise, returns null (go).
     */
    public readonly playCard = async (playedCards: Card[]) => {
        const playableCardIndex = this.handCards.findIndex((card) => {
            return checkPlayable(card, playedCards);
        });
        if (playableCardIndex === -1) {
            return null;
        }
        const [removedCard] = this.handCards.splice(playableCardIndex, 1);
        return removedCard;
    };

    public readonly setHandCards = (cards: Card[]) => {
        this.handCards = [...cards];
    };

    public readonly getHandCards = () => {
        return [...this.handCards];
    };
}

/**
 * @brief Player that takes decisions from stdin.
 */
export class HumanPlayer extends Player {
    /**
     * @brief Allows a card to be selected from the player's hand.
     */
    public readonly selectCard = async () => {
        console.log('Enter a number to select the corresponding card.');
        console.log(`${this.handCards}`);
        [true, false].forEach((arrow) => {
            this.handCards.forEach((card, i) => {
                if (card.rank == 10) {
                    process.stdout.write(' ');
                }
                process.stdout.write(`${arrow ? '^' : `${i + 1}`}  `);
            });
            console.log();
        });
        const index: number = (
            await prompts({
                type: 'number',
                name: 'index',
                message: 'Choose',
                validate: (value: number) => {
                    return (
                        (value >= 1 && value <= this.handCards.length) ||
                        `Please enter a number between 1 and ${this.handCards.length}`
                    );
                },
            })
        ).index;
        const selectedCard = this.handCards[index - 1];
        this.handCards.splice(index - 1, 1);
        return selectedCard;
    };

    public readonly throwToCrib = async (dealer: boolean) => {
        const cardsToThrow: Card[] = [];
        console.log(
            `Throw 2 cards into your ${dealer ? '' : "opponent's"} crib.`,
        );
        for (const cardName of ['first', 'second']) {
            console.log();
            console.log(`Throw the ${cardName} card.`);
            cardsToThrow.push(await this.selectCard());
        }
        console.log();
        return cardsToThrow;
    };

    public readonly playCard = async (
        playedCards: Card[],
    ): Promise<Card | null> => {
        for (const card of this.handCards) {
            if (checkPlayable(card, playedCards)) {
                console.log();
                console.log('Play a card.');
                const selectedCard = await this.selectCard();
                if (checkPlayable(selectedCard, playedCards)) {
                    return selectedCard;
                }
                console.log(
                    `${selectedCard} is not playable. Select a different card.`,
                );
                this.handCards.push(selectedCard);
                return this.playCard(playedCards);
            }
        }
        console.log('No playable cards left.');
        return null;
    };
}